File: BOJ/24_DFS와BFS/2178_미로탐색.cpp
#include <cstdio>
#include <iostream>
#include <queue>

using namespace std;

/*
    숫자가 공백없이 한줄로 입력이 되서 들어오면 scanf("%1d") 로 입력받으면 숫자 하나씩 입력 가능.
*/


int N,M;
int board[101][101];
int dx[4] = {-1,1,0,0}, dy[4] = {0,0,-1,1};

void bfs() {

    queue<pair<int,int> > q;
    q.push(make_pair(0,0));

    while(!q.empty()) {
        int cx = q.front().first;
        int cy = q.front().second;
        q.pop();
        for(int i=0; i<4; i++) {
            int nx = cx + dx[i];
            int ny = cy + dy[i];
            if(0 <= nx && nx < N && 0<= ny && ny < M && board[nx][ny] == 1) {
                q.push(make_pair(nx,ny));
                board[nx][ny] = board[cx][cy] + 1;
            }
        }
    }
}


int main() {

    scanf("%d %d",&N,&M);
    for(int i=0; i<N; i++)
        for(int j=0; j<M; j++) 
            scanf("%1d",&board[i][j]);

    bfs();

    cout << board[N-1][M-1] << endl;

    return 0;
}

File: BOJ/24_DFS와BFS/2178_미로탐색.py
# 얘도 하나씩 구분지어서 받으려면 board = [ ... for _ ] 가 아닌
# board.append(list(input())) 으로 해줘야 함 -> 구분지어서 받을 수 있음.
# 그리고 더해줄 수 있게 board[0][0] = 1 로 해줌. 
# 이하 같음.
def bfs() :
    queue = [(0,0)]
    while queue :
        cx,cy = queue.pop(0)
        for (dx,dy) in zip(dxs,dys) :
            nx = cx + dx
            ny = cy + dy
            if 0 <= nx < N and 0 <= ny < M and board[nx][ny] == '1' :
                queue.append((nx,ny))
                board[nx][ny] = board[cx][cy] + 1

N,M = map(int,input().split())
#board = [input() for _ in range(N)]
board = []
for _ in range(N) :
    board.append(list(input()))
dxs = [1,0,-1,0]
dys = [0,1,0,-1]
board[0][0] = 1
bfs()
print(board)